### PR TITLE
Implement updating of version from git tag, and retagging from @jspaaks

### PR DESCRIPTION
**Pull request by alexlancaster**
This workaround adds support for getting the version info from the tag used in the release and Zenodo DOI into the `CITATION.cff` and `.zenodo.json` files into the repo, and then retagging the final produce before uploading the zipball to zenodo.

### CODE REVIEW

**Comment by alexlancaster** [comment]
hi @jspaaks, I have ported and merged this back upstream to my `zenodo_upload` branch.  I am testing it now.  A couple of questions:

1. will this work if the branch being released from is not the default one (i.e. `main`)?  I ask because I'm testing this from `zenodo_upload` branch, and it looks like it's running the updating files part, but the results aren't being committed back to the `zenodo_branch` (nor any branch, including `main`) on the github repo? see: https://github.com/alexlancaster/pypop/actions/runs/6332973168/job/17200378707
2. when you were testing this on the main zenodo, how did you ensure that the record wouldn't be published?  I'm guess this is because you had this previously "dummy" record: https://zenodo.org/record/8383808 and that you set the `collection` ID to that record, and then set the `published: false`, which puts it in draft mode and doesn't actually create a real DOI, is that right?  is ok to create an empty "test" collection like this, and if so, how did you do it?

**Comment by alexlancaster** [comment]
Hi @jspaaks OK for (1), above, I did some serious git sleuthing and I think I finally got this to get effect I want in the github action:

https://github.com/alexlancaster/pypop/blob/e61afef80350c50dab37c350d1cbacfcf7cadca5/.github/workflows/build_wheels.yml#L141

you can see the output here:

https://github.com/alexlancaster/pypop/actions/runs/6334732402/job/17204812048

however it has the side-effect of turning the published release, back into "draft" status, see here:

https://github.com/alexlancaster/pypop/releases/tag/untagged-ebb78ab0a5ecfecd41fa

 (I guess because the tag disappears and reappears with a a new SHA).  I'm worried now that if I publish it, it will start the whole cycle again.  So, I'm sort of stumped.  Not sure what the main zenodraft action does here - does it try to move the tag to include the change to the zenodo DOI?

**Comment by alexlancaster** [comment]
and I clicked publish - and yep - sure enough it does restart the build etc.

**Comment by alexlancaster** [comment]
or maybe the  new tag creation part needs to be done by zenodraft itself at the end, rather than after the version update part?  I'm guessing that must be it.

 I can't test that right now due to sandbox being down, and not having a "test" collection as I mentioned for issue (2), above.

**Comment by jspaaks** [comment]
The zenodraft action should delete the release + tag the user makes via GitHub's frontend, and then creates a new release with a new tag pointing to the commit that has any updates that occurred in the workflow. So my guess is, you dont need

https://github.com/alexlancaster/pypop/blob/e61afef80350c50dab37c350d1cbacfcf7cadca5/.github/workflows/build_wheels.yml#L152C1-L155C34

As to creating the test deposition on Zenodo, the text below is copied from https://github.com/zenodraft/action/issues/22#issuecomment-1738625575

I first tried using the Sandbox, but (1) their site looked different and (2) it appeared to have problems, both of which I attributed to the migration work that was going on. So I opted for using the regular Zenodo instance (that looked like normal) by having `sandbox: false` and different credentials ($ZENODO_ACCESS_TOKEN instead of $ZENODO_SANDBOX_ACCESS_TOKEN). Because I was just doing some testing, I set `publish: false` and I also used a new collection by using the Zenodo form (Upload button > New upload > Add a nonempty file > Save but not publish the deposition will give you a new draft record number. Decrement by 1 to get the collection/concept number). It's not strictly necessary but this way it's easier to avoid clutter. So with that setup you get depositions on Zenodo, but they remain unfinalized until you click the button in Zenodo's form. You can see which unfinalized depositions you have by logging in to Zenodo and then clicking the "Upload" button near the top of the page, that will have the overview.

So basically you had the right idea in https://github.com/alexlancaster/pypop/pull/129#issuecomment-1738290370' s bullet 2 :)

**Comment by alexlancaster** [comment]
I think I'm getting close!

> The zenodraft action should delete the release + tag the user makes via GitHub's frontend, and then creates a new release with a new tag pointing to the commit that has any updates that occurred in the workflow. So my guess is, you dont need
> 
> https://github.com/alexlancaster/pypop/blob/e61afef80350c50dab37c350d1cbacfcf7cadca5/.github/workflows/build_wheels.yml#L152C1-L155C34

Removed this.

> As to creating the test deposition on Zenodo, the text below is copied from [zenodraft/action#22 (comment)](https://github.com/zenodraft/action/issues/22#issuecomment-1738625575)
> 
> I first tried using the Sandbox, but (1) their site looked different and (2) it appeared to have problems, both of which I attributed to the migration work that was going on. So I opted for using the regular Zenodo instance (that looked like normal) by having `sandbox: false` and different credentials ($ZENODO_ACCESS_TOKEN instead of $ZENODO_SANDBOX_ACCESS_TOKEN). Because I was just doing some testing, I set `publish: false` and I also used a new collection by using the Zenodo form (Upload button > New upload > Add a nonempty file > Save but not publish the deposition will give you a new draft record number. Decrement by 1 to get the collection/concept number). It's not strictly necessary but this way it's easier to avoid clutter. 

OK I did that, and created a test deposit with just a license file (it's not published, but it has the URL)

https://zenodo.org/deposit/8387712

I took the number 8387712, and decremented by 1, to give 8387711, put that into the action here:

https://github.com/alexlancaster/pypop/blob/fe58de2983326c4a45c82727711bd0c5ee4e198d/.github/workflows/build_wheels.yml#L166

then rebuild for a release - it gets further, but it still fails on creating a deposition with:

```
  creating a new, empty versioned deposition in existing collection...
  checking that the collection_id value resolves to a concept, not a version...
  getting id of the latest version in the collection...
  Error: Cannot read properties of undefined (reading 'split')
```

https://github.com/alexlancaster/pypop/actions/runs/6342677312/job/17228950578#step:6:197

> So with that setup you get depositions on Zenodo, but they remain unfinalized until you click the button in Zenodo's form. You can see which unfinalized depositions you have by logging in to Zenodo and then clicking the "Upload" button near the top of the page, that will have the overview.

Do I need to actually publish that initial empty deposition so that the collection is "seen"? I notice that yours is public.

**Comment by alexlancaster** [comment]
@jspaaks there's also another problem I have to solve.  The wheels and dist via CI should also build *after* the retagging, otherwise the source and wheels that go to pypi will not match the tag.  So I now make the `publish_zenodo` job *precede* the wheel and dist building.  I modified the action to skip the zenodo step for all but releases and dropped the zenodraft part, but kept the retagging, but now I have two problems when I run the job:

1. although the zenodo retagging seems to work the resulting dist doesn't have the changes (even though the retagging happens *before* the wheels process starts).  I downloaded the artifacts and unzipped the tar, but it doesn't contain the .cff and .zenodo.json changes
2. for some reason the upload_to_gh release doesn't run (even though the wheels and sdist are generated)

e.g. see: https://github.com/alexlancaster/pypop/actions/runs/6356852773  (the wheels have the right 0.9.40 version number, but the contents contain 0.9.39 in the CITATION.cff and .zenodo.json

see the current wheels file: https://github.com/alexlancaster/pypop/blob/zenodo_upload/.github/workflows/build_wheels.yml

**Comment by alexlancaster** [comment]
In the example build above, looking a bit closer at the log for the checkout for the `build_sdist` it looks like it's still pointing to the SHA of the v0.9.40 tag *before* the retagging:

```
 From https://github.com/alexlancaster/pypop
   * [new ref]         872e4d4cd45d790e91da01fe4c0bc72084deb980 -> v0.9.40
```

https://github.com/alexlancaster/pypop/actions/runs/6356852773/job/17267107109#step:2:54

whereas it should be 913213e763118735a1f3592dc062d33c65fe5e26 (created in the previous step), which should have been done here:

```
To https://github.com/alexlancaster/pypop
 * [new tag]         v0.9.40 -> v0.9.40
```

https://github.com/alexlancaster/pypop/actions/runs/6356852773/job/17267100375#step:5:73

Is there an additional step I need to make sure that the git repo in the next step has the updated v0.9.40 tag?

**Comment by alexlancaster** [comment]
OK, it looks like I figured out my answer to (1), above.  It seems that the `action/checkout@v4` by default uses the SHA of the original commit that triggered the workflow, rather than the tag name that was used in the release event.  But you can use the `with:` to override the default checkout, So I've modified workflow:

https://github.com/alexlancaster/pypop/blob/9ce7b48951e96834d808e378d0911d4d63175752/.github/workflows/build_wheels.yml#L69

```
          ref: ${{ (github.event_name == 'release' && github.event.action == 'published') && github.ref_name || '' }}
```

basically if a published event happened, we checkout the **new** tag, otherwise empty string `''` defaults back to the standard behaviour.  It's a bit hacky, but it seems to work. More elegant solutions are welcome.

Now I'm working on why (2) is not working.

**Comment by alexlancaster** [comment]
OK, I now think I've also figured out (2).  It seems a job will be skipped by default if any previous steps have been skipped (even intentionally).  The ugly workaround is to add `always() && !cancelled() && !failure()` after each `if:` conditional, just as was one with the `build_sdist` job.  

https://github.com/alexlancaster/pypop/blob/c1449fa25912b373b37daf6014ac452ca928bf5a/.github/workflows/build_wheels.yml#L84

So it not only looks at the parents job status, but the grandparents too, which is completely unintuitive.  But their ya go.

**Comment by alexlancaster** [comment]
So now the only outstanding piece getting the zenodraft working.  I can partially emulate the workflow without getting a zenodo DOI, except that once the tag is moved - the release notes gets separated from the new tag, so you get two releases, one with the notes, and the other with the uploaded files. e..g for 0.9.48:

https://github.com/alexlancaster/pypop/releases/tag/v0.9.48

and

https://github.com/alexlancaster/pypop/releases/tag/untagged-869ae140383bdd407fba

@jspaaks is there any github action tor command-line hat can bring them back into alignment until such time as the zenodraft is working (or I can test on sandbox and not worry about creating DOIs, because they will be fake? I'm assuming that zenodraft must realign the tags and the release notes internally somehow?

**Comment by alexlancaster** [comment]
> @jspaaks is there any github action tor command-line hat can bring them back into alignment until such time as the zenodraft is working (or I can test on sandbox and not worry about creating DOIs, because they will be fake?) I'm assuming that zenodraft must realign the tags and the release notes internally somehow?

Answering my own question, I was able to cobble together from some existing actions, copying the data from the old release, deleting it, and then recreating a new with the same data and the updated tag here:

https://github.com/alexlancaster/pypop/blob/f2be2d9d784c36ba5badd44f13d6fb0c33da80d1/.github/workflows/build_wheels.yml#L85

it seems to update properly and when the wheels and .tar.gz are ready, the `upload_gh_release` job appears to put them in the right place, e.g. here:

https://github.com/alexlancaster/pypop/actions/runs/6360589556

the only thing I lose is the title on the workflow run (presumably because the release id is deleted).
